File: Java/Combo.java
import java.util.*;
import java.io.*;

public class Combo {
	
	static int n, k;
	static long mod = (long)(1e9+7);
	static long[][] choose;
	static long[] fact, inv_fact;
	
	public static void main(String[] args) {
		
		choose = new long[n+1][k+1];
		build();
		
		fact = new long[n+1]; 
		inv_fact = new long[n+1];
		fact[0] = inv_fact[0] = 1;
		for (int i = 1; i < fact.length; i++) {
			fact[i] = fact[i-1] * i;
			fact[i] %= mod;
			inv_fact[i] = pow(fact[i], mod-2, mod);
		}
	}
	
	public static void build() {
		for (int i = 0; i < choose.length; i++) {
			for (int j = 0; j < choose[0].length && j <= i; j++) {
				if (j == 0 || j == i) choose[i][j] = 1;
				else {
					choose[i][j] = choose[i-1][j-1] + choose[i-1][j];
					choose[i][j] %= mod;
				}
			}
		}
	}
	
		// top! / bottom! (top - bottom)!
	public static long choose(int top, int bottom) {
		return fact[top] * inv_fact[bottom] % mod * inv_fact[top - bottom] % mod;
	}
	
    public static long pow(long a, long b, long m) {
    	long ans = 1;
    	while (b > 0) {
    		if ((b & 1) == 1) {
    			ans *= a;
    			ans %= m;
    		}
    		a *= a;
    		a %= m;
    		b >>= 1;
    	}
    	return ans;
    }
}

File: C++/Combo.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <string>
#include <string.h>
#include <array>
#include <unordered_map>
#include <unordered_set>
#include <iomanip>
using namespace std;
typedef long long ll;
typedef long double ld;
#define pb push_back

const int MaxN = 1e4+10;
const int MaxK = 1e4+10;
const ll mod = 1e9+7;
ll choose[MaxN][MaxK];

ll fact[MaxN], invFact[MaxN];

void build() {
    for (int i = 0; i < MaxN; i++) {
        for (int j = 0; j < MaxK && j <= i; j++) {
            if (j == 0 || j == i) {
                choose[i][j] = 1;
            }
            else {
                choose[i][j] = choose[i-1][j-1] + choose[i-1][j];
                choose[i][j] %= mod;
            }
        }
    }
}

    // top! / bottom! (top - bottom)!
ll C(int top, int bottom) {
    return fact[top] * invFact[bottom] % mod * invFact[top - bottom] % mod;
}

ll pow(ll a, ll b, ll m) {
    ll ans = 1;
    while (b > 0) {
        if ((b & 1) == 1) {
            ans *= a;
            ans %= m;
        }
        a *= a;
        a %= m;
        b >>= 1;
    }
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    fact[0] = invFact[0] = 1;
    for (int i=1; i<MaxN; i++) {
        fact[i] = fact[i-1] * i;
        fact[i] %= mod;
        invFact[i] = pow(fact[i], mod-2, mod);
    }
    
}